README.md
# Table of Contents

1. [description](#description)

2. [data-format](#data-format)

3. [batch-training](#batch-training)


# description

Package:      areas/neural/systems/cascor/

Name:         CASCOR

Summary:      Lisp and C implementations of Cascade Correlation

Version:      C 1.1 (31-MAY-94); Lisp 1.0 ()

Description:

   This directory contains C and Lisp implementations of the Cascade
   Correlation Algorithm.

   The Lisp code was written by Scott Fahlman, then ported to C by
   Scott Crowder. More recent C versions have been re-engineered by Matt
   White.  Features have been added to allow data sets from the CMU
   learning benchmark database to run on this system.

Requires:     C, Lisp

Ports:

Origin:       ftp.cs.cmu.edu:/afs/cs.cmu.edu/project/connect/code/
              as the files cascor-v1.1.shar and cascor1.lisp
              and in /afs/cs.cmu.edu/project/connect/tr/ as cascor-tr.ps.Z

Copying:      Public Domain

Updated:      Thu Sep  1 19:18:52 1994

CD-ROM:       Prime Time Freeware for AI, Issue 1-1

Bug Reports:  neural-bench@cs.cmu.edu

Mailing List:

Author(s):    Scott Fahlman <sef+@cs.cmu.edu>
              R. Scott Crowder <crowders@tcplink.nrel.gov>
              Matt White <mwhite+@cmu.edu>

Contact:      neural-bench@cs.cmu.edu

Keywords:

   Neural Networks, Cascade Correlation, C!Code, Lisp!Code, CMU,
   Public Domain, Authors!Fahlman, Authors!Crowder, Authors!White,
   Machine Learning!Neural Networks, CASCOR

Contains:    ???

See Also:

   areas/neural/bench/

References:

   Scott E. Fahlman and Christian Lebiere, "The Cascade-Correlation
   Learning Architecture", in D. S. Touretzky, editor, "Advances in
   Neural Information Processing Systems 2", Morgan Kaufmann, 1990.  A
   somewhat longer version is available as CMU Computer Science Tech
   Report CMU-CS-90-100, a copy of which is included in the distribution.

   Marcus Hoehfeld and Scott E. Fahlman (1992) "Learning with Limited
   Numerical Precision Using the Cascade-Correlation Learning Algorithm"
   in IEEE Transactions on Neural Networks, Vol. 3, no. 4, July 1992, pp.
   602-611.


# data-format

CMU Neural Network Learning Benchmark Database
Data File Format

Maintainer:  neural-bench@cs.cmu.edu

  This file contains a description of the standard format for data files in the
CMU learning benchmark collection.  These files are a supplement to the
benchmark description files that comprise the CMU Benchmark Collection.  Data
files are associated to their appropriate description file with a '.data'
extension to the file name of the benchmark description.


  SEGMENTS
  --------
  Each data set is composed of two to four segments.  The first segment
is always the $SETUP segment.  The $SETUP segment is immediately followed by
the $TRAIN segment.  There are also optional $VALIDATION and $TEST segments.

  $SETUP
  ======
    The $SETUP segment describes the requirements of the network to the
  program.  Included in this segment is information on how the program should
  read the actual data segments as well as what type of inputs and outputs are
  required.  All lines in the $SETUP segment should end in a ';'.

    PROTOCOL: {IO, SEQUENCE};
      The protocol parameter tells the program how to read the data sets
    included in the file.  In an IO mapping, each vector of input and output
    values is a seperate training case, independant of all others.  The
    network's output depends only on the current inputs.  In a SEQUENCE
    mapping, the input/output vectors are presented in sequential order.  The
    output may depend on earlier inputs as well as the current input vector.

    OFFSET: <n>;
      This appears only in SEQUENCE mappings.  It is the number of input
    vectors to read before an output should be produced.  For most problems,
    this will be set to '0'.

    INPUTS: <n>;
      This is the number of items in an input vector.  However, since some
    data types, such as enumerations, may require more than one input unit to
    represent, the actual number of input units may be greater.

    OUTPUTS: <n>;
      Similar to INPUTS, this specifies outputs instead of inputs.  Again,
    due to some data types requiring more than one unit to represent, there
    may be a disparity between this number and the actual number of output
    nodes in the network.

    IN [n]: < CONT {Domain}, BINARY, ENUM {list} >;
      Each input must have an explicit entry describing it.  This entry
    contains the node number (n) and a data type.  Available types are:
    CONTinuous, BINARY, and ENUMerated.  Continuous inputs are floating
    point numbers with a specified domain (where all numbers are guaranteed to
    fall).  Binary inputs have either a value of '+' or '-'.  An enumerated
    input is one of a list specified within '{}'.

    OUT [n]: < CONT {CoDomain}, BINARY, ENUM {list} >;
      Each output must also have an explicit entry describing it.  This entry
    contains the node number (n) and a data type.  A CONT output is a
    floating point output which is guaranteed to fall within a specified
    CoDomain.  BINARY outputs should have either a positive, '+', or a
    negative, '-', value.  An ENUMerated output is one of the specified list.

    NOTE -  While listing node types, no fields are acceptable.  In other
    words, the definition 'IN [1..13]: BINARY' or 'OUT [1]: ENUM {A..Z}'
    would NOT be legal.


  $TRAIN, $VALIDATION, $TEST
  --------------------------
    These segments contain the actual training, validation and testing data to
  be used by the network.  The validation and testing segments are optional.
  Entries into one of these segments should have the form:

	<input 1>, <input 2>, <etc>  =>  <output 1>, <output 2>, <etc>;

    In SEQUENCE data sets, there may also be '<>' delimiters.  These specify
  the end of a sub-sequence.  Data sets on opposite sides of one of these
  delimiters should not affect eachother.  The sequence-end delimiters do NOT
  require a semicolon, as do not segment headers.
    In data sets with an offset, there should be no arrow and no outputs for
  the first (n) inputs.  Simply end the list of inputs with a semicolon.


  COMMENTS
  --------
    While no actual comment card is specified, any text occuring on a line
  after a semicolon should NOT be parsed.  Therefore it is possible to
  comment a data file by starting a line with a semicolon.


# batch-training

From sef@sef-pmax.slisp.cs.cmu.edu Wed Mar  2 18:52:51 EST 1994
Article: 14911 of comp.ai.neural-nets
Xref: glinda.oz.cs.cmu.edu comp.ai.neural-nets:14911
Newsgroups: comp.ai.neural-nets
Path: honeydew.srv.cs.cmu.edu!news
From: sef@sef-pmax.slisp.cs.cmu.edu
Subject: Re: online cascade-correlation
Message-ID: <CLMvwu.1Ku.3@cs.cmu.edu>
Sender: news@cs.cmu.edu (Usenet News System)
Nntp-Posting-Host: sef-pmax.slisp.cs.cmu.edu
Organization: School of Computer Science, Carnegie Mellon
Date: Tue, 22 Feb 1994 16:17:17 GMT
Lines: 47


    From: B35@vm.urz.uni-heidelberg.de (Matthijs Kadijk)

    In a recent posting by Scott E. Fahlman, he wrote that some people
    have addapted cascor for online (i.s.o. batch) learning. I would be
    interested to try this out the learning problem I'm corrently working
    on.

I don't have code for this, but what you need to do is the following:

1. Replace the quickprop weight update with simple gradient descent, with
or without momentum according to taste.

2. Rip out the cacheing mechanism, since it no longer makes sense to cache
the network's results for a whole epoch.

3. Modify the quiescence tests to keep some kind of running average and to
stop when there is no net improvement for a long time.  Note that in online
updating, the error or correlation score does not smoothly approach an
asymptote, but continues to bounce around as new samples arrive.

4. Optional: In the online case, it may be desirable to run output and
candidate training concurrently.  Just make sure that candidate training
continues for some time after the output weights have finally been frozen.

5. If you are running the net for performance and training at the same
time, introduce new hidden units with a very small initial output weight
(of the proper sign) so that you don't create too much of a bump in the
network's performance.

By the way, I have seen a few claims that Cascor runs better if Quickprop
is replaced by simple gradient descent, even in the batch-training case.  I
don't believe these claims, and that has not been my experience.  I will
admit that it can be a bit tricky to adjust the epsilon parameter in
Quickprop to get best performance, however.

-- Scott

===========================================================================
Scott E. Fahlman			Internet:  sef+@cs.cmu.edu
Senior Research Scientist		Phone:     412 268-2575
School of Computer Science              Fax:       412 681-5739
Carnegie Mellon University		Latitude:  40:26:33 N
5000 Forbes Avenue			Longitude: 79:56:48 W
Pittsburgh, PA 15213
===========================================================================


CHANGELOG.md
# Changelog

# Todo
- none

# v2023.4.1-beta
- initial version


cascor.h
/*	Cascade Correlation Learning Algorithm

	v1.1
	Matt White  (mwhite+@cmu.edu)
        5/31/93
*/

#ifndef CASCOR
#define CASCOR

/*	Include Files	*/

#include "tools.h"
#include "parse.h"


/*	Data Enumeration Definitions	*/

typedef enum  {		/*  This enumeration describes the four modes the  */
  TRAINING,		/* network can enter into during training.	   */
  TIMEOUT,		/* TRAINING means that the network is still        */
  STAGNANT,		/* training.  TIMEOUT means that the network       */
  WIN			/* reached the maximum number of epochs for the    */
}  status_t;		/* last training cycle.  STAGNANT meand that,	   */
			/* during the last training cycle,  a point was    */
/* reached where performance ceased to improve by an appreciable amount.   */
/* WIN simply means that network performance reached an acceptable level   */
/* during the last training cycle.					   */


typedef enum  {		/*  error_t describes the two methods for 	*/
  BITS,			/* calculating error.  Either the program can   */
  INDEX			/* count the number of responses that are	*/
}  error_t;		/* wrong, or it can sum overall error among all */
			/* the outputs.					*/


typedef enum  {		/*  An enumeration of the various unit types.	*/
  VARIED,		/* Each unit type has a specific activation	*/
  UNINITIALIZED,	/* function associated with it.  VARIED		*/
  SIGMOID,		/* indicates a mix of the other unit types	*/
  ASIGMOID,
  GAUSSIAN,
  VARSIGMOID,
  LINEAR
}  unit_type;


/*	Data Structure Definitions	*/

/*	NET_DATA -  Holds weights and values for the network as a whole
*/

typedef struct  {
  unit_type  *unitTypes;  /*  Unit types for each unit in the network       */
  float      *values,	  /*  Activation value of each unit in the network  */
            **weights;	  /*  Value of weights connecting the units	  */
}  net_data;


/*	OUTPUT_DATA -  Holds values, weights and training information about the
	outputs of the network.
*/

typedef struct  {
  float      shrinkFactor,	/*  Shrink factor to be used by quickprop    */
             scaledEpsilon,	/*  Output epsilon, scaled by the number of  */
				/* training points being used.		     */
             *values,		/*  Activation value of each output	     */
             **weights,		/*  Weights connecting each output to inputs */
				/* and hidden units			     */
             **deltas,		/*  Change made in this particular weight    */
				/* last epoch.  Used in quickprop.	     */
             **slopes,		/*  Partial derivitive of error with respect */
				/* to the weight.			     */
             **pSlopes;		/*  Previous partial derivitive		     */
  unit_type  *types;		/*  Unit type of each output		     */
}  output_data;


/*	CAND_DATA -  This structure contains information to the training and
	selecting of new units to be added to the network.
*/

typedef struct  {
  int        best;		/*  Number of the current best unit	    */
  float      shrinkFactor,	/*  Shrink factor for candidates	    */
             bestScore,		/*  Correlation value for the current best  */
				/* candidate				    */
             *values,		/*  Activation value of each candidate	    */
             *sumVals,		/*  Summed activation over all training	    */
				/* examples.				    */
             **weights,		/*  Weight values between candidates and    */
				/* inputs and hidden units		    */
             **corr,		/*  Correlation between candidates and each */
				/* output in the network		    */
             **pCorr,		/*  Previous correlation values		    */
             **deltas,		/*  Change made in this weight last epoch   */
             **slopes,		/*  Slope of the correlation to weight	    */
				/* curve.  Used in quickprop.		    */
             **pSlopes;		/*  Previous values of the slope	    */
  unit_type  *types;		/*  Type of each individual candidate	    */
}  cand_data;


/*	ERROR_DATA -  This structure contains information used to determine how
	well the network is doing.  Anything to do with the network's error.
*/

typedef struct  {
  int      bits;		/*  Number of bits incorrect		     */
  float    stdDev,		/*  Standard deviation of all training	     */
				/* points being used			     */
           index,		/*  Error index.  Used to determine whether  */
				/* a network is done training.		     */
           indexThresh,		/*  How good the error index has to get      */
				/* before training is finished		     */
           trueErr,		/*  Sum squared raw error		     */
           sumSqErr,		/*  Sum squared error prime		     */
           scoreThresh,		/*  How close an output has to be to its     */
				/* correct value to be considered correct.   */
				/* Used in BITS error measure.		     */
           *sumErr,		/*  The sum of the errors for all training   */
				/* points, by output			     */
           *errors;		/*  Error at each output		     */
  error_t  measure;		/*  Which method to use to calculate error   */
}  error_data;


/*	NODE_PARMS -  This structure holds parameter values for both the
	candidate as well as outputs.
*/

typedef struct  {
  int    epochs,		/*  Number of epochs to train before timeout */
         patience;		/*  How long to continue training without    */
				/* noticable improvement before calling the  */
				/* cycle stagnant			     */
  float  sigMax,		/*  Maximum value for VARSIGMOID units to    */
				/* stretch to 			             */
         sigMin,		/*  Minimum value for VARSIGMOID units to    */
				/* stretch to				     */
         epsilon,		/*  Epsilon, or learning rate to use	     */
         decay,			/*  Slope decay per epoch.  Controls         */
				/* floating point overflow.		     */
	 mu,			/*  Maximum growth factor.  Quickprop	     */
         changeThresh;		/*  Determines what is considered noticable  */
				/* change for stagnation purposes	     */
}  node_parms;


/*	PARM_DATA -  This structure contains user set parameters used by the
	program, except those that pertain to error.  Everything that should
	be user settable, with the exception of error parameters, is located
	withing this structure.
*/

typedef struct  {
  int         Ntrials,		/*  Number of trials to run                  */
              maxNewUnits,	/*  Maximum number of new units to add to    */
			        /* the network				     */
              winRadius,	/*  Radius of window to use on sequence      */
				/* data sets				     */
              valPatience;	/*  Number of cycles to continue without     */
				/* improvement on validation set performance */
				/* before stopping trial		     */
  float       weightRange,	/*  Range of initial random weights	     */
              weightMult,	/*  Multiplier for generating weights 	     */
				/* between a new unit and the outputs	     */
              sigPrimeOffset,	/*  Used to eliminate offset in sigmoid      */
				/* activation functions			     */
              bias;		/*  Activation level for the bias unit	     */
  char        *dataFile,	/*  Name of the data file being used	     */
              *weightFile;	/*  Name of the weight file being used       */
  boolean     parseInBinary,	/*  Parse enumerated network inputs as	     */
				/* binary values as opposed to unary values  */
              parseOutBinary,	/*  Parse enumerated network outputs as	     */
				/* binary values instead of unary values     */
              validate,		/*  Perform a validation epoch every cycle   */
              test,		/*  Perform a test epoch every trial	     */
              useCache,		/*  Cache activation values and errors	     */
              saveWeights;	/*  Save the weights at the end of each      */
				/* trial				     */
  unit_type   candNewType;	/*  Type of candidates to use in cand pool   */
  node_parms  out,		/*  Output unit parameters		     */
              cand;		/*  Candidate unit parameters		     */
}  parm_data;


/*	ALT_DATA_T -  Holds information about alternate data sets in the data
	file.  Namely this means the validation and test sets.  This structure
	is essential to being able to validate and test the network.
*/

typedef struct  {
  int       Npts,		/*  Number of points, total, in this set     */
            Nsegs,		/*  Number of segment markers in this set    */
            NoutVals,		/*  Number of total output values in the set */
            bestPass;		/*  Validation only, which cycle was best?   */
  float     stdDev,		/*  Standard deviation of the set	     */
            bestScore,		/*  True error value at the end of the best  */
				/* cycle				     */
            **bestOutConn;	/*  Value of each weight connecting inputs   */
				/* and hidden units to the outputs during    */
				/* the 'best' cycle			     */
  data_set  data;		/*  The actual data points		     */
}  alt_data_t;


/*	CACHE_DATA -  Since internal weights do not change in cascor, it is
	possible to store the internal activations and errors from epoch to
	epoch, not having to recalculate every time.  Both values and errors
	are stored for every training point, making a cache expensive, memory-
	wise.
*/

typedef struct  {
  float  **values,	/*  Cached activation values of hidden units	*/
         **errors;	/*  Cached error values of hidden units		*/
}  cache_data;


/*	RUN_RES_T -  Stores the cumulative statistics of the run for the final
	end report.
*/

typedef struct  {
  int     Nvictories,		/*  Number of victories won		     */
          Nepochs,		/*  Average number of epochs until victory   */
          errorBits,		/*  Average number of bits wrong	     */
	  Nunits;		/*  Average number of units in the completed */
				/* network				     */
  float   crossingsSec,		/*  Average number of connection crossings   */
				/* per second				     */
          percentCorrect,	/*  Percent of bits correct		     */
          runTime,		/*  Length of seconds, of the run	     */
          errorIndex,		/*  Average error index			     */
          trueError,		/*  Average true error			     */
          sumSqError;		/*  Average sum squared error		     */
}  run_res_t;


/*	Constant Declarations	*/

#define CONNX				/*  Turn connection crossing	    */
					/* statistics ON		    */

#define MAX_PATH 	40		/*  Maximum pathname length	    */
#define N_EXT		3		/*  Total number of extensions      */
#define DATA_EXT	".data"		/*  Data file extension		    */
#define CONFIG_EXT	".cfg"		/*  Configuration file extension    */
#define WEIGHT_EXT	".net"		/*  Save file extension		    */
#define HELP_FILE	"cascor.hlp"	/*  User help file		    */

#define BIN_POS        0.5		/*  Value for binary '+'	    */
#define BIN_NEG        -0.5		/*  Value for binary '-'	    */

#define DEF_VALIDATE	FALSE		/*  Default value for parm.validate */
#define DEF_TEST	TRUE		/*  Default value for parm.test     */

#define VERSION		"1.1"			    /*  Version number	*/
#define RELEASE_DATE	"5/31/94"		    /*  Release date	*/
#define CONTACT		"neural-bench@cs.cmu.edu"   /*  Site contact at CMU */

/*	Macro to calculate error index	*/

#define ERROR_INDEX( TE, sDev, num )  ( sqrt( TE / num ) / sDev )


#endif
